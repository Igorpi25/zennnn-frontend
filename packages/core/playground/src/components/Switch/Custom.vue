<template>
  <Switch
    label="Red"
    input-class="peer-checked:bg-green-500 dark:peer-checked:bg-green-500 peer-focus:ring-green-400"
  >
    <span class="ml-2">Green color</span>
  </Switch>
</template>
