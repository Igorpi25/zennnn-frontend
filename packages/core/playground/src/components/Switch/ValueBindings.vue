<template>
  <Switch
    v-model="model"
    true-value="on"
    false-value="off"
  >
    <span class="ml-2">Value: {{ model }}</span>
  </Switch>
</template>

<script setup>
import { ref } from 'vue'

const model = ref('off')
</script>
