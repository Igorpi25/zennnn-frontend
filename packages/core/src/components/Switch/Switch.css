@layer components {
  .switch__control {
    @apply flex items-center relative;
  }
  .switch--show-details .switch__control {
    @apply pb-2;
  }
  .switch__control__input {
    width: 45px;
    height: 24px;
    @apply relative;
  }
  .switch--small .switch__control__input {
    width: 30px;
    height: 16px;
  }

  .switch__track,
  .switch__thumb {
    @apply pointer-events-none;
  }

  .switch__track {
    transition: .3s cubic-bezier(.25,.8,.5,1);
    @apply rounded-full absolute bg-light-gray-400 dark:bg-gray-900 w-full h-full;
  }
  .switch--focused .switch__track {
    @apply ring ring-blue-400;
  }
  .switch__input:focus:not(:focus-visible) + .switch__track {
    box-shadow: none;
    outline: 0;
  }
  .switch--active .switch__track {
    @apply bg-blue-500;
  }
  .switch--disabled .switch__track {
    @apply bg-light-gray-400 dark:bg-gray-900;
  }

  .switch__thumb {
    height: 18px;
    width: 18px;
    left: 3px;
    top: calc(50% - 9px);
    transition: .3s cubic-bezier(.25,.8,.5,1);
    @apply bg-gray-100 dark:bg-gray-200 rounded-full relative;
  }
  .switch--active .switch__thumb {
    transform: translateX(21px);
    @apply bg-white;
  }
  .switch--small .switch__thumb {
    height: 12px;
    width: 12px;
    left: 2px;
    top: calc(50% - 6px);
  }
  .switch--active.switch--small .switch__thumb {
    transform: translateX(14px);
  }

  .switch__input {
    @apply absolute opacity-0 select-none w-full h-full;
  }

  .switch__label {
    @apply select-none leading-tight;
  }
  .switch--small .switch__label {
    @apply leading-none;
  }

  .switch__label,
  .switch__input {
    @apply cursor-pointer;
  }
  .switch--disabled .switch__label,
  .switch--disabled .switch__input {
    @apply cursor-not-allowed;
  }

  .switch--disabled {
    @apply opacity-50;
  }
}
