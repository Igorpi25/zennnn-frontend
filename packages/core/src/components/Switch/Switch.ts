import { h, ref, computed, watch, defineComponent, mergeProps } from 'vue'
import { deepEqual } from 'vue-supp'
import { useInputProps, useInput } from '../../composables/useInput'
import {
  useInputValidationProps,
  useInputValidation,
} from '../../composables/useInputValidation'
import uid from '../../utils/uid'

import type { PropType } from 'vue'

export default defineComponent({
  name: 'Switch',

  props: {
    ...useInputProps(),
    ...useInputValidationProps(),
    modelValue: {
      type: [Boolean, String, Number],
      default: false,
    },
    value: {
      type: [Boolean, String, Number],
      default: false,
    },
    valueComparator: {
      type: Function,
      default: deepEqual,
    },
    trueValue: null,
    falseValue: null,
    small: Boolean,
    controlClass: String,
    onChange: Function as PropType<(val: any) => void>,
  },

  slots: ['label', 'default'],

  emits: ['update:modelValue', 'update:error', 'change'],

  setup(props, { slots, emit }) {
    const id: string = uid('switch-')
    const rootElement = ref<HTMLElement>()

    const { inputElement, internalValue, isFocused, inputData, genLabel } =
      useInput(props, {
        slots,
        id,
      })

    const { showDetails, isDisabled, isReadonly, isInteractive, genMessages } =
      useInputValidation(props, { emit, id, internalValue, isFocused })

    const isActive = computed(() => {
      const value = props.value
      const input = internalValue.value

      if (props.trueValue === undefined || props.falseValue === undefined) {
        return value ? props.valueComparator(value, input) : Boolean(input)
      }

      return props.valueComparator(input, props.trueValue)
    })

    watch(
      () => props.modelValue,
      (val) => {
        internalValue.value = val
      }
    )

    watch(internalValue, (val) => {
      emit('update:modelValue', val)
    })

    function onIconClick() {
      if (isDisabled.value || isReadonly.value) return
      ;(inputElement.value as HTMLInputElement).click()
    }

    function onChange() {
      if (!isInteractive.value) return

      const value = props.value
      let input = internalValue.value

      if (props.trueValue !== undefined && props.falseValue !== undefined) {
        input = props.valueComparator(input, props.trueValue)
          ? props.falseValue
          : props.trueValue
      } else if (value) {
        input = props.valueComparator(input, value) ? null : value
      } else {
        input = !input
      }

      internalValue.value = input
      emit('change', internalValue.value)
    }

    function onKeydown(e: KeyboardEvent) {
      if (
        ((e.key === 'ArrowLeft' || e.key === 'Left') && isActive.value) ||
        ((e.key === 'ArrowRight' || e.key === 'Right') && !isActive.value)
      )
        onChange()
    }

    function genSwitchLabel() {
      const children = slots.default?.()
      if (!children) return undefined
      return h(
        'label',
        {
          for: id,
          class: 'switch__label',
        },
        children
      )
    }

    function genSwitchInput() {
      return h(
        'input',
        mergeProps(inputData.value, {
          id,
          class: 'peer sr-only',
          value: props.value,
          checked: isActive.value,
          type: 'checkbox',
          role: 'checkbox',
          readonly: isReadonly.value,
          disabled: isDisabled.value,
          'aria-checked': isActive.value,
          onChange: onChange,
          onKeydown: onKeydown,
        })
      )
    }

    function genSwitchIcon() {
      return h('div', {
        class: [
          'switch__icon',
          {
            'switch__icon--small': props.small,
          },
          props.inputClass,
        ],
        onClick: onIconClick,
      })
    }

    function genControl() {
      return h(
        'div',
        {
          class: [
            'switch__control',
            {
              'switch__control--show-details': showDetails.value,
            },
            props.controlClass,
          ],
        },
        [genSwitchInput(), genSwitchIcon(), genSwitchLabel()]
      )
    }

    return () => {
      return h(
        'div',
        {
          ref: rootElement,
          class: {
            switch: true,
            'switch--disabled': isDisabled.value,
            'switch--readonly': isReadonly.value,
          },
        },
        [genLabel(), genControl(), genMessages()]
      )
    }
  },
})
