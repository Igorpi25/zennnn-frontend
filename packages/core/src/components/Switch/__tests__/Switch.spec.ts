import { mount, VueWrapper } from '@vue/test-utils'

import Switch from '../Switch'

jest.mock('../../../utils/uid')

describe('Switch.ts', () => {
  let mountFunction: (options?: Record<string, unknown>) => VueWrapper<any>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Switch, {
        ...options,
      })
    }
  })

  // Render

  it('should render component and match snapshot', () => {
    const wrapper = mountFunction({
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with disabled prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        disabled: true,
      },
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with name, required, readonly props and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        name: 'simple-input',
        required: true,
        readonly: true,
      },
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render checked component and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        modelValue: true,
      },
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render focused component and match snapshot', async () => {
    const wrapper = mountFunction({
      slots: {
        default: () => 'Label',
      },
    })

    await wrapper.get('input').trigger('focus')

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with label prop and match snapshot', async () => {
    const wrapper = mountFunction({
      props: {
        label: 'Input label',
      },
      slots: {
        default: () => 'Switch label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  // ---

  it('should emit true/false when input/label clicked', async () => {
    const wrapper = mountFunction({
      slots: {
        default: () => 'Label',
      },
    })

    await wrapper.get('input').trigger('click')
    await wrapper.get('label').trigger('click')

    expect(wrapper.emitted()).toHaveProperty('change')
    expect(wrapper.emitted().change[0]).toEqual([true])
    expect(wrapper.emitted().change[1]).toEqual([false])
  })

  it('should emit specified value and return null on toggled off when clicked', async () => {
    const wrapper = mountFunction({
      props: {
        value: 'Test',
      },
    })

    const input = wrapper.get('input')
    await input.trigger('click')
    await input.trigger('click')

    expect(wrapper.emitted().change[0]).toEqual(['Test'])
    expect(wrapper.emitted().change[1]).toEqual([null])
  })

  it('should work with custom true- and false-value', async () => {
    const wrapper = mountFunction({
      props: {
        trueValue: 'on',
        falseValue: 'off',
      },
    })

    const input = wrapper.get('input')
    await input.trigger('click')
    await input.trigger('click')

    expect(wrapper.emitted().change[0]).toEqual(['on'])
    expect(wrapper.emitted().change[1]).toEqual(['off'])
  })

  it('should toggle on arrow left and right keydown', async () => {
    const wrapper = mountFunction()

    const input = wrapper.get('input')
    await input.trigger('keydown', { key: 'ArrowRight' })
    await input.trigger('keydown', { key: 'ArrowLeft' })

    expect(wrapper.emitted().change[0]).toEqual([true])
    expect(wrapper.emitted().change[1]).toEqual([false])
  })
})
