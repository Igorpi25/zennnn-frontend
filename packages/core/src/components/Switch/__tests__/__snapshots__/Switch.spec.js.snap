// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch.js should render checked component and match snapshot 1`] = `
<div class="switch switch--active">
  <!---->
  <div class="switch__control">
    <div class="switch__control__input"><input id="switch-1" class="switch__input" type="checkbox" value="false" role="checkbox" aria-checked="true">
      <div class="switch__track"></div>
      <div class="switch__thumb"></div>
    </div><label for="switch-1" class="switch__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Switch.js should render component and match snapshot 1`] = `
<div class="switch">
  <!---->
  <div class="switch__control">
    <div class="switch__control__input"><input id="switch-1" class="switch__input" type="checkbox" value="false" role="checkbox" aria-checked="false">
      <div class="switch__track"></div>
      <div class="switch__thumb"></div>
    </div><label for="switch-1" class="switch__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Switch.js should render component with disabled prop and match snapshot 1`] = `
<div class="switch switch--disabled">
  <!---->
  <div class="switch__control">
    <div class="switch__control__input"><input disabled="" id="switch-1" class="switch__input" type="checkbox" value="false" role="checkbox" aria-checked="false">
      <div class="switch__track"></div>
      <div class="switch__thumb"></div>
    </div><label for="switch-1" class="switch__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Switch.js should render component with label prop and match snapshot 1`] = `
<div class="switch"><label id="label-1" for="switch-1" aria-hidden="false" class="label">
    <div class="label__wrapper">
      <div class="label__content__wrapper">
        <div class="label__content">Input label</div>
      </div>
      <!---->
    </div>
  </label>
  <div class="switch__control">
    <div class="switch__control__input"><input id="switch-1" class="switch__input" type="checkbox" value="false" role="checkbox" aria-checked="false">
      <div class="switch__track"></div>
      <div class="switch__thumb"></div>
    </div><label for="switch-1" class="switch__label">Switch label</label>
  </div>
  <!---->
</div>
`;

exports[`Switch.js should render component with name, required, readonly props and match snapshot 1`] = `
<div class="switch switch--readonly">
  <!---->
  <div class="switch__control">
    <div class="switch__control__input"><input name="simple-input" required="" readonly="" id="switch-1" class="switch__input" type="checkbox" value="false" role="checkbox" aria-checked="false">
      <div class="switch__track"></div>
      <div class="switch__thumb"></div>
    </div><label for="switch-1" class="switch__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Switch.js should render focused component and match snapshot 1`] = `
<div class="switch switch--focused">
  <!---->
  <div class="switch__control">
    <div class="switch__control__input"><input id="switch-1" class="switch__input" type="checkbox" value="false" role="checkbox" aria-checked="false">
      <div class="switch__track"></div>
      <div class="switch__thumb"></div>
    </div><label for="switch-1" class="switch__label">Label</label>
  </div>
  <!---->
</div>
`;
